util/prism_utils.ts

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 the Deno authors. All rights reserved. MIT license.
+// Taken from https://github.com/FormidableLabs/prism-react-renderer/blob/master/src/utils/normalizeTokens.js
+
+// Empty lines need to contain a single empty token, denoted with { empty: true }
+function normalizeEmptyLines(line: any[]) {
+  if (line.length === 0) {
+    line.push({
+      types: ["plain"],
+      content: "\n",
+      empty: true,
+    });
+  } else if (line.length === 1 && line[0].content === "") {
+    line[0].content = "\n";
+    line[0].empty = true;
+  }
+}
+
+function appendTypes(types: string[], add: string[] | string): string[] {
+  const typesSize = types.length;
+  if (typesSize > 0 && types[typesSize - 1] === add) {
+    return types;
+  }
+
+  return types.concat(add);
+}
+
+const newlineRe = /\r\n|\r|\n/;
+
+// Takes an array of Prism's tokens and groups them by line, turning plain
+// strings into tokens as well. Tokens can become recursive in some cases,
+// which means that their types are concatenated. Plain-string tokens however
+// are always of type "plain".
+// This is not recursive to avoid exceeding the call-stack limit, since it's unclear
+// how nested Prism's tokens can become
+export function normalizeTokens(tokens: Array<any | string>): any[][] {
```

Comment:
1. It's unfortunate this is needed.

2. Use jsdoc 

3. Consider adding a test for this

Reply:
1. yea i do agree, but prism doesnt provide a way to get tokens as lines
2. copied as is from the original code, but sure
3. sure